Add custom `stopping_criteria` and  `logits_processor` to `generate`

lvwerra opened the pull request:
# What does this PR do?

This PR are continues the work and discussions from #12219 with a fresh start.

It integrates the custom `stopping_criteria` and `logits_processor` with the following logic: 
- the default `stopping_criteria`/`logits_processor` are created from the arguments and model's config 
- if additional, custom `stopping_criteria` and `logits_processor` are passed to `generate`, they are compared with the default list
   - if there is an overlap between the two lists (e.g. a `MaxLengthCriteria` in both lists) an error is thrown
   - if there is no overlap the two lists are merged 

Fixes #12118

Commit:
add custom `stopping_criteria` and `logits_processor` to `generate`

Commit:
add tests for custom `stopping_criteria` and `logits_processor`

Commit:
fix typo in RAG

Narsil reviewed (approved):
Looks good and clean.

Only suggestion would be making it slightly more obvious in the docstring that those arguments are intended for power users.

Commit:
address reviewer comments

lvwerra commented:
@Narsil thanks for your feedback and catching that copy-paste error! I added something to the docstring and fixed the error.

@patrickvonplaten any comments?

Commit:
Merge branch 'master' into custom-criteria-logits-processor-generate

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
+    def _merge_criteria_processor_list(
+        self, default_list: StoppingCriteriaList, custom_list: StoppingCriteriaList
+    ) -> StoppingCriteriaList:
+        if custom_list is None:
```
How can the custom_list be None? I think it can only be empty no? Wouldn't it be better to do:

```python
if len(custom_list) == 0:
    return default_list
```

lvwerra replied:
The default in `generate` is `None` but I can change it to an empty list if you think that is more consistent.

lvwerra replied:
An empty `StoppingCriteriaList` and `LogitsProcessorList` to be more precise.

patrickvonplaten commented on src/transformers/models/rag/modeling_rag.py:
```diff
@@ -1469,6 +1473,14 @@ def generate(
                 argument is useful for constrained generation conditioned on the prefix, as described in
                 `Autoregressive Entity Retrieval <https://arxiv.org/abs/2010.00904>`__.
+            logits_processor (:obj:`LogitsProcessorList`, `optional`):
+                 Custom logits processors that complement the default logits processors built from arguments and a
```
I like the doc-string!

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
+            for custom in custom_list:
+                if type(custom) is type(default):
+                    raise ValueError(
+                        f"A custom stopping criteria or logits processor of type {type(custom)} was passed to `generate` which is already created with an argument or model config."
```
```suggestion
                        f"A custom stopping criteria or logits processor of type {type(custom)} was passed "
                        "to `generate` which is already created with an argument or model config."
```

patrickvonplaten replied:
Also, since this error won't be easy to solve for the user can we try to give her/him as much guidance as possible? we could a) Figure out if it's a `stopping criteria` or a `logits processor` from the type. And then b) IMO we should display as much information as possible. E.g.
```
object_type = "stopping criteria" if isinstance(custom, StoppingCriteria) else "logits processor"
f"A custom {object_type} of type {type(custom)} with values {custom} has been passed to `generate`, but it has already been created with the values {default}. {default} has been created by passing the corresponding arguments to generate or by the config's default values."

patrickvonplaten replied:
Something like this maybe, there is probably a better way to phrase it

patrickvonplaten replied:
Actually what we could do in a second step is to enforce that the __init__ names of the processors have to correspond 1-to-1 to the inputs of `generate()` and then we can actually pass the arguments that need to set to `None` in generate to allow this. Maybe a bit overkill in a first step, but we could do this if we encounter issues about it

lvwerra replied:
I refined the error message to include a little bit more detail. Let me know if you think it is clearer now.

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
+        return criteria
+
+    def _merge_criteria_processor_list(
+        self, default_list: StoppingCriteriaList, custom_list: StoppingCriteriaList
```
```suggestion
        self, default_list: List[Any], custom_list: List[Any]
```

Narsil replied:
They are not  of type `List`. `Union[LogitsProcessorList, StoppingCriteriaList]` is the correct thing (you can define a new type too to make it less verbose maybe)

```python
GenerateList = Union[LogitsProcessorList, StoppingCriteriaList]

    def _merge(... default_list: GenerateList) -> GenerateList:
       ....
```

lvwerra replied:
I have added `Union[LogitsProcessorList, StoppingCriteriaList]`. Should be more careful with type hints in the future. I chose the verbose option because it is quite readable that way.

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
+
+    def _merge_criteria_processor_list(
+        self, default_list: StoppingCriteriaList, custom_list: StoppingCriteriaList
+    ) -> StoppingCriteriaList:
```
```suggestion
    ) -> List[Any]:
```

lvwerra replied:
See above discussion.

Commit:
improve custom logits processor/stopping criteria error message

Commit:
fix types in merge function signature

Commit:
change default for custom list from `None` to empty list

Commit:
Merge branch 'master' into custom-criteria-logits-processor-generate

Commit:
fix rag generate

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -712,15 +714,36 @@ def _get_logits_processor(
+                if type(custom) is type(default):
+                    object_type = "stopping criteria" if isinstance(custom, StoppingCriteria) else "logits processor"
+                    raise ValueError(
+                        f"A custom {object_type} of type {type(custom)} with values {custom} has been passed to `generate`, but it has already been created with the values {default}. {default} has been created by passing the corresponding arguments to generate or by the model's config default values. If you just want to change the default values of {object_type} consider passing them as arguments ...
```
@sgugger would be super happy if we can split this into multiple lines for readability ;-)
```suggestion
                        f"A custom {object_type} of type {type(custom)} with values {custom} has been passed to `generate`, "
                        f"but it has already been created with the values {default}. {default} has been created by passing the "
                        "corresponding arguments to generate or by the model's config default values. "
                        f"If you just want to change the default values of {object_type} consider passing them as arguments "
                        f"to `generate` instead of using a custom {object_type}."
```

patrickvonplaten reviewed (approved):
Great work! Thanks a lot for the feature. 

Feel free to merge whenever :-)

lvwerra committed:
add string split suggestion

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>
